plugins/keyboard-navigation/src/navigation.js

```diff
@@ -0,0 +1,377 @@
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace to get the marker
+   *     from.
+   * @return {Blockly.Marker} The marker created for keyboard navigation.
+   * @protected
+   */
+  getMarker(workspace) {
+    return workspace.getMarker(this.MARKER_NAME);
+  }
+
+  /**
+   * Adds all event listeners and cursors to the flyout that are needed for
+   * keyboard navigation to work.
+   * @param {!Blockly.IFlyout} flyout The flyout to add a cursor and change
+   *     listeners to.
+   * @protected
+   */
+  addFlyout(flyout) {
+    const flyoutWorkspace = flyout.getWorkspace();
+    flyoutWorkspace.addChangeListener(this.flyoutChangeWrapper);
+    const FlyoutCursorClass = Blockly.registry.getClass(
+        cursorRegistrationType, cursorRegistrationName);
+    flyoutWorkspace.getMarkerManager().setCursor(new FlyoutCursorClass());
+  }
+
+  /**
+   * Removes all change listeners from the flyout that are needed for
+   * keyboard navigation to work.
+   * @param {!Blockly.IFlyout} flyout The flyout to add a cursor and event
+   *     listeners to.
+   * @protected
+   */
+  removeFlyout(flyout) {
+    const flyoutWorkspace = flyout.getWorkspace();
+    flyoutWorkspace.removeChangeListener(this.flyoutChangeWrapper);
+  }
+
+  /**
+   * Updates the state of keyboard navigation and the position of the cursor
+   * based on workspace events.
+   * @param {!Blockly.Events.Abstract} e The Blockly event to process.
+   * @protected
+   */
+  workspaceChangeListener(e) {
+    const workspace = Blockly.Workspace.getById(e.workspaceId);
+    if (!workspace || !workspace.keyboardAccessibilityMode) {
+      return;
+    }
+    switch (e.type) {
+      case Blockly.Events.DELETE:
+        this.handleBlockDeleteByDrag(workspace, e);
+        break;
+      case Blockly.Events.BLOCK_CHANGE:
+        if (e.element === 'mutation') {
+          this.handleBlockMutation(
+              workspace, /** @type {Blockly.Events.BlockChange} */ (e));
+        }
+        break;
+      case Blockly.Events.CLICK:
+        this.handleWorkspaceClick(
+            workspace, /** @type {Blockly.Events.Click} */ (e));
+        break;
+      case Blockly.Events.TOOLBOX_ITEM_SELECT:
+        this.handleToolboxCategoryClick(
+            workspace, /** @type {Blockly.Events.ToolboxItemSelect} */ (e));
+        break;
+      case Blockly.Events.BLOCK_CREATE:
+        this.handleBlockCreate(workspace, e);
+    }
+  }
+
+  /**
+   * Updates the state of keyboard navigation and the position of the cursor
+   * based on events emitted from the flyout's workspace.
+   * @param {!Blockly.Events.Abstract} e The Blockly event to process.
+   * @protected
+   */
+  flyoutChangeListener(e) {
+    const flyoutWorkspace = Blockly.Workspace.getById(e.workspaceId);
+    const mainWorkspace = flyoutWorkspace.targetWorkspace;
+    const flyout = mainWorkspace.getFlyout();
+
+    if (mainWorkspace && mainWorkspace.keyboardAccessibilityMode &&
+        !flyout.autoClose) {
```

Comment:
This somewhat confuses me. do we not need to do anything if the flyout autocloses? or on the other hand, it looks like this only works on "simple" flyouts which i didn't think could close. So either way I'm a bit confused on this condition (no change needed, just asking for my own understanding)

Reply:
So I think this is for the case when the toolbox has a flyout and autoclose = false. `workspace.getFlyout()` returns either the flyout on the toolbox, or the simple flyout on the workspace.

Reply:
Ok, added a comment. Lmk if that helps!